renku/ui/service/controllers/cache_files_upload.py

```diff
@@ -56,6 +58,70 @@ def context(self):
         """Controller operation context."""
         return self.response_builder
 
+    def process_upload(self):
+        """Process an upload."""
+        if self.response_builder.get("chunked_id", None) is None:
+            return self.process_file()
+
+        return self.process_chunked_upload()
+
+    def process_chunked_upload(self):
+        """Process upload done in chunks."""
+        if self.response_builder["total_size"] > MAX_CONTENT_LENGTH:
+            if MAX_CONTENT_LENGTH > 524288000:
+                max_size = bytes_to_unit(MAX_CONTENT_LENGTH, "gb")
+                max_size_str = f"{max_size} gb"
+            else:
+                max_size = bytes_to_unit(MAX_CONTENT_LENGTH, "mb")
+                max_size_str = f"{max_size} mb"
+            raise UserUploadTooLargeError(maximum_size=max_size_str, exception=None)
+
+        chunked_id = self.response_builder["chunked_id"]
+
+        user_cache_dir = CACHE_UPLOADS_PATH / self.user.user_id
+        chunks_dir = user_cache_dir / chunked_id
+        chunks_dir.mkdir(exist_ok=True, parents=True)
+
+        current_chunk = self.response_builder["chunk_index"]
+        total_chunks = self.response_builder["chunk_count"]
+
+        file_path = chunks_dir / str(current_chunk)
+        relative_path = file_path.relative_to(CACHE_UPLOADS_PATH / self.user.user_id / chunked_id)
+
+        self.file.save(str(file_path))
+
+        with self.cache.model_db.lock(f"chunked_upload_{self.user.user_id}_{chunked_id}"):
+            self.cache.set_file_chunk(
+                self.user,
+                {
+                    "chunked_id": chunked_id,
+                    "file_name": str(current_chunk),
+                    "relative_path": str(relative_path),
+                },
+            )
+            completed = len(list(self.cache.get_chunks(self.user, chunked_id))) == total_chunks
+
+        if not completed:
+            return {}
+
+        target_file_path = user_cache_dir / self.file.filename
+
+        if target_file_path.exists():
+            if self.response_builder.get("override_existing", False):
+                target_file_path.unlink()
+            else:
+                raise IntermittentFileExistsError(file_name=self.file.filename)
+
+        with open(target_file_path, "wb") as f:
+            for file_number in range(total_chunks):
+                f.write((chunks_dir / str(file_number)).read_bytes())
```

Comment:
TIL about `shutil.copyfileobj` (https://stackoverflow.com/a/27077437). Should we use it there to avoid possible use of large buffers.